marlowe/test/Spec/Marlowe/Common.hs

```diff
@@ -6,54 +6,332 @@ module Spec.Marlowe.Common where
+shrinkChoiceId _ = []
+
+
+valueIdGen :: Gen ValueId
+valueIdGen = oneof [ return "alpha"
+                   , return "beta"
+                   ]
+
+
+shrinkValueId :: ValueId -> [ValueId]
+shrinkValueId "beta"  = ["alpha"]
+shrinkValueId "alpha" = []
+shrinkValueId _       = []
+
+
+rationalGen :: Gen P.Rational
+rationalGen = do
+    a <- simpleIntegerGen
+    b <- positiveAmount
+    return $ a % b
+
+
+valueGenSized :: Int -> Gen (Value Observation)
+valueGenSized s
+  | s > 0 = oneof [ AvailableMoney <$> accountIdGen <*> tokenGen
+                  , Constant <$> simpleIntegerGen
+                  , NegValue <$> valueGenSized (s - 1)
+                  , AddValue <$> valueGenSized (s `quot` 2) <*> valueGenSized (s `quot` 2)
+                  , SubValue <$> valueGenSized (s `quot` 2) <*> valueGenSized (s `quot` 2)
+                  , ChoiceValue <$> choiceIdGen <*> valueGenSized (s - 1)
+                  , Scale <$> rationalGen <*> valueGenSized (s `quot` 2)
+                  , Cond  <$> observationGenSized s
+                          <*> valueGenSized (s `quot` 2)
+                          <*> valueGenSized (s `quot` 2)
+                  , return SlotIntervalStart
+                  , return SlotIntervalEnd
+                  , UseValue <$> valueIdGen
+                  ]
+  | otherwise = oneof [ AvailableMoney <$> accountIdGen <*> tokenGen
+                      , Constant <$> simpleIntegerGen
+                      , return SlotIntervalStart
+                      , return SlotIntervalEnd
+                      , UseValue <$> valueIdGen
+                      ]
+
+
+valueGen ::  Gen (Value Observation)
+valueGen = sized valueGenSized
+
+
+shrinkValue :: (Value Observation) -> [(Value Observation)]
+shrinkValue value = case value of
+    Constant x -> [Constant y | y <- shrinkSimpleInteger x]
+    SlotIntervalStart -> [Constant 0]
+    SlotIntervalEnd -> [Constant 0, SlotIntervalStart]
+    AvailableMoney accId tok -> Constant 0 : ([AvailableMoney x tok | x <- shrinkAccountId accId]
+               ++ [AvailableMoney accId y | y <- shrinkToken tok])
+    UseValue valId -> Constant 0 : [UseValue x | x <- shrinkValueId valId]
+    ChoiceValue choId val -> Constant 0 : val : ([ChoiceValue x val | x <- shrinkChoiceId choId]
+                   ++ [ChoiceValue choId y | y <- shrinkValue val])
+    NegValue val -> Constant 0 : val : [NegValue x | x <- shrinkValue val]
+    AddValue val1 val2 -> Constant 0 : val1 : val2 : ([AddValue x val2 | x <- shrinkValue val1]
+                         ++ [AddValue val1 y | y <- shrinkValue val2])
+    SubValue val1 val2 -> Constant 0 : val1 : val2 : ([SubValue x val2 | x <- shrinkValue val1]
+                         ++ [SubValue val1 y | y <- shrinkValue val2])
+    Scale r val -> Constant 0 : val : [Scale r v | v <- shrinkValue val]
+    Cond b val1 val2 -> Constant 0 : val1 : val2 : ([Cond b x val2 | x <- shrinkValue val1]
+                         ++ [Cond b val1 y | y <- shrinkValue val2])
```

Comment:
Could you add also shrinks for the observation in `Cond`? You can get a template from `shrinkObservation`

Reply:
It didn't seem to terminate, so I removed it. Probably my mistake, will try again.